### PR TITLE
Add MapArray constructors and doc example

### LINKED ISSUE #4385: Add doc example of constructing a MapArray
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
People on a discord channel were recently confused about creating `MapArray`s 

The current documentation https://docs.rs/arrow/41.0.0/arrow/array/struct.MapArray.html does not give an example of how to create `MapArrays`

**Describe the solution you'd like**
A clear example of how to construct a `MapArray` linked from the `MapArray` docs

**Describe alternatives you've considered**

**Additional context**
@tustvold suggests "using the builder" which I think means https://docs.rs/arrow/41.0.0/arrow/array/struct.MapBuilder.html

Here is the example code for constructing it:

```rust
fn build_map_array(list: Vec<HashMap<String, proto::Value>>) -> Result<Arc<MapArray>> {
    let mut entry_offset = vec![];
    let mut keys = vec![];
    let mut values = vec![];

    let mut offset = 0;
    for kv in list {
        entry_offset.push(offset as u64);
        offset += kv.len();

        for (key, value) in kv {
            keys.push(key);
            values.push(format!("{}:{}", value.type_name(), value));
        }
    }
    println!("offset list: {:?}", entry_offset);

    let map_data = ArrayDataBuilder::new(datatype_map())
        .len(3)
        .add_buffer(Buffer::from_vec(entry_offset))
        .add_child_data(
            StructArray::from(vec![
                (
                    Arc::new(Field::new("keys", DataType::Utf8, false)),
                    Arc::new(StringArray::from(keys)) as ArrayRef,
                ),
                (
                    Arc::new(Field::new("values", DataType::Utf8, false)),
                    Arc::new(StringArray::from(values)) as ArrayRef,
                ),
            ])
            .into_data(),
        )
        .build()?;

    Ok(Arc::new(MapArray::from(map_data)))
}
``` 
(which I am not sure is right)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4385

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/array/list_array.rs [hunk 1]
```diff
@@ -103,7 +104,7 @@ impl<OffsetSize: OffsetSizeTrait> GenericListArray<OffsetSize> {
         let len = offsets.len() - 1; // Offsets guaranteed to not be empty
         let end_offset = offsets.last().unwrap().as_usize();
         // don't need to check other values of `offsets` because they are checked
-        // during construction of `OffsetsbBuffer`
+        // during construction of `OffsetsBuffer`
```
Comment (viirya) [inline arrow-array/src/array/list_array.rs:107]:
```suggestion
        // during construction of `OffsetBuffer`
```

File: arrow-array/src/array/map_array.rs [hunk 2]
```diff
@@ -40,6 +40,106 @@ pub struct MapArray {
...
+            if n.len() != len {
+                return Err(ArrowError::InvalidArgumentError(format!(
+                    "Incorrect length of null buffer for MapArray, expected {len} got {}",
+                    n.len(),
+                )));
+            }
+        }
+        if field.is_nullable() || entries.null_count() != 0 {
+            return Err(ArrowError::InvalidArgumentError(
+                "MapArray entries cannot contain be null".to_string(),
```
Comment (viirya) [inline arrow-array/src/array/map_array.rs:83]:
```suggestion
                "MapArray entries cannot contain null".to_string(),
```

File: arrow-array/src/array/map_array.rs [hunk 3]
```diff
@@ -40,6 +40,106 @@ pub struct MapArray {
...
+        field: FieldRef,
+        offsets: OffsetBuffer<i32>,
+        entries: StructArray,
+        nulls: Option<NullBuffer>,
+        ordered: bool,
+    ) -> Result<Self, ArrowError> {
+        let len = offsets.len() - 1; // Offsets guaranteed to not be empty
+        let end_offset = offsets.last().unwrap().as_usize();
+        // don't need to check other values of `offsets` because they are checked
+        // during construction of `OffsetsBuffer`
```
Comment (viirya) [inline arrow-array/src/array/map_array.rs:65]:
```suggestion
        // during construction of `OffsetBuffer`
```

File: arrow-array/src/array/map_array.rs [hunk 4]
```diff
@@ -19,16 +19,18 @@ use crate::array::{get_offsets, print_long_array};
...
 /// An array of key-value maps
 ///
 /// Keys should always be non-null, but values can be null.
 ///
-/// [MapArray] is physically a [crate::array::ListArray] that has a
-/// [StructArray] with 2 child fields.
+/// [`MapArray`] is physically a [`ListArray`] of key values pairs stored as an `entries`
+/// [`StructArray`] with 2 child fields.
+///
+/// See [`MapBuilder`](crate::builder::MapBuilder) for how to construct a [`MapArray`]
```
Comment (tustvold) [inline arrow-array/src/array/map_array.rs:33]: I didn't see a compelling reason to duplicate the example already on MapBuilder

**Review by alamb — APPROVED** [review]

**Pull request merged**